src/types/bytes.rs

```diff
@@ -26,6 +27,87 @@ impl PyBytes {
         unsafe { py.from_owned_ptr(ffi::PyBytes_FromStringAndSize(ptr, len)) }
     }
 
+    /// Creates a new Python bytestring object and a mutable slice to its underlying buffer.
+    /// The bytestring is uninitialised and must not be read.
+    ///
+    /// Panics if out of memory.
+    #[inline]
+    unsafe fn new_with_uninit_impl(
+        py: Python<'_>,
+        len: usize,
+    ) -> (&mut [MaybeUninit<u8>], &PyBytes) {
+        let length = len as ffi::Py_ssize_t;
+        let pyptr = ffi::PyBytes_FromStringAndSize(std::ptr::null(), length);
+        // Iff pyptr is null, py.from_owned_ptr(pyptr) will panic
+        let pybytes = py.from_owned_ptr(pyptr);
+        let buffer = ffi::PyBytes_AsString(pyptr) as *mut MaybeUninit<u8>;
+        debug_assert!(!buffer.is_null());
+        let slice = std::slice::from_raw_parts_mut(buffer, len);
+        (slice, pybytes)
+    }
+
+    /// Creates a new Python bytestring object.
+    /// The bytestring is zero-initialised and can be read inside `init`.
+    /// The `init` closure can further initialise the bytestring.
+    ///
+    /// Panics if out of memory.
+    ///
+    /// # Example
+    /// ```
+    /// use pyo3::{prelude::*, types::PyBytes};
+    /// Python::with_gil(|py| -> PyResult<()> {
+    ///     let py_bytes = PyBytes::new_with(py, 10, |bytes: &mut [u8]| {
+    ///         bytes.copy_from_slice(b"Hello Rust");
+    ///     });
+    ///     let bytes: &[u8] = FromPyObject::extract(py_bytes)?;
+    ///     assert_eq!(bytes, b"Hello Rust");
+    ///     Ok(())
+    /// });
+    /// ```
+    pub fn new_with<F: Fn(&mut [u8])>(py: Python<'_>, len: usize, init: F) -> &PyBytes {
+        let (slice, pybytes) = unsafe { Self::new_with_uninit_impl(py, len) };
+        slice.iter_mut().for_each(|x| *x = MaybeUninit::zeroed());
```

Comment:
If you choose to remove `new_with_uninit`, you may find `std::ptr::write_bytes` a nice alternative to making a `MaybeUninit` slice.

Reply:
Thanks for the suggestion - I’ll keep that in mind.